logprep/util/rule_loader.py

```diff
@@ -0,0 +1,257 @@
+        the specified directory.
+
+        Returns
+        -------
+        list of dict
+          A list of rule definitions created from the rule files in the directory recursively.
+        """
+        rule_files = self.files_by_extensions(self.source)
+        rule_definitions = (
+            FileRuleLoader(str(file), self.processor_name).rule_definitions for file in rule_files
+        )
+        return list(itertools.chain(*rule_definitions))
+
+
+class FileRuleLoader(RuleLoader):
+    """FileRuleLoader is responsible for loading rules from a file.
+    The file can contain multiple rules defined in either JSON or YAML format.
+
+    Parameters
+    ----------
+    source : str
+      The path to the file containing the rule definitions.
+    processor_name : str
+      The name of the processor that the rules belong to.
+    """
+
+    @property
+    def rules(self) -> List[Rule]:
+        """Retrieve a list of rules.
+        This method processes the rule definitions and returns a list of Rule objects.
+        It ensures that the source attribute is a string before proceeding.
+
+        Returns
+        -------
+        List[Rule]
+          A list of Rule objects generated from the rule definitions.
+
+        Raises
+        ------
+        TypeError
+          If the source attribute is not a string.
+        """
+
+        if not isinstance(self.source, str):
+            raise TypeError(f"Expected a string, got {type(self.source)}")
+        try:
+            return list(
+                itertools.chain(
+                    *[
+                        DictRuleLoader(rule_definition, self.processor_name).rules
+                        for rule_definition in self.rule_definitions
+                    ]
+                )
+            )
+        except ValueError as error:
+            logger.warning("Loading rules from %s failed: %s -> Skipping.", self.source, error)
+        return []
+
+    @property
+    def rule_definitions(self) -> List[Dict]:
+        rule_definitions: List[Dict] | Dict = []
+        if not isinstance(self.source, str):
+            raise TypeError(f"Expected a string, got {type(self.source)}")
+        if self.source.endswith(".yaml") or self.source.endswith(".yml"):
+            rule_definitions = GetterFactory.from_string(self.source).get_yaml()
+        elif self.source.endswith(".json"):
+            rule_definitions = GetterFactory.from_string(self.source).get_json()
+        if not isinstance(rule_definitions, list):
+            rule_definitions = [rule_definitions]
+        return rule_definitions
+
+
+class DictRuleLoader(RuleLoader):
+    """DictRuleLoader is responsible for loading a rule from a dictionary.
+    The dictionary should reflect a valid rule definition.
+
+    Parameters
+    ----------
+    source : dict
+      The dictionary containing the rule definition.
+    processor_name : str
+      The name of the processor that the rules belong to.
+    """
+
```

Comment:
Shouln'd there be an `__init__` method like in `DirectoryRuleLoader`, since `source` must be a `dict`?

Reply:
not needed as in the other child classes. I wrote one to easily fix mypy errors. but the other way around this is more robust. Validate the type with an exception.